DREAM/Optimization/tasdreamOptimizationTests.cpp

```diff
@@ -0,0 +1,194 @@
+    f(best_swarm_point, best_swarm_value_vec);
+    pass = pass & (fabs(best_swarm_value_vec[0] - optimal_val) <= tol);
+    std::vector<bool> init_vector = state.getStateVector();
+    pass = pass & (init_vector[3]);
+
+    // Make sure subsequent runs do not make any strange modifications.
+    ParticleSwarm(f, 1, inside, state, 0.5, 2, 2, get_rand);
+    f(best_swarm_point, best_swarm_value_vec);
+    pass = pass & (fabs(best_swarm_value_vec[0] - optimal_val) <= tol);
+    init_vector = state.getStateVector();
+    pass = pass & (init_vector[3]);
+
+    // TasOptimization::ParticleSwarmState::clearCache().
+    state.clearCache();
+    init_vector = state.getStateVector();
+    pass = pass & (init_vector[0] and init_vector[1] and init_vector[2] and !init_vector[3]);
+
+    return pass;
+}
+
+// Unit tests for TasOptimization::ParticleSwarm on multiple objective functions.
+bool testParticleSwarm(double tol=1e-2) {
+    bool pass = true;
+
+    // l1 norm over the domain [-5, 2] ^ 6.
+    int num_dimensions = 6;
+    int num_particles = 100;
+    int iterations = 200;
+    std::vector<double> lower(num_dimensions, -5);
+    std::vector<double> upper(num_dimensions, 2);
+    TasOptimization::ObjectiveFunctionSingle l1_single =
+            [](const std::vector<double> &x)->double {
+                double sum = 0;
+                for (auto xi : x) sum += fabs(xi);
+                return sum;};
+    TasOptimization::ObjectiveFunction l1 = TasOptimization::makeObjectiveFunction(num_dimensions, l1_single);
+    TasOptimization::ParticleSwarmState state(num_dimensions, num_particles);
+    state.initializeParticlesInsideBox(lower, upper);
+    pass = pass & (testParticleSwarmSingle(l1, state, TasDREAM::hypercube(lower, upper), iterations, 0, tol));
+
+
+    // Six hump-camel function over the domain [-3, 3] x [-2, 2].
+    num_dimensions = 2;
+    num_particles = 50;
+    iterations = 20;
+    lower = {-3, -2};
+    upper = {3, 2};
+    TasOptimization::ObjectiveFunctionSingle shc_single =
+            [](const std::vector<double> &x)->double {
+                return (4 - 2.1 * x[0]*x[0] + x[0]*x[0]*x[0]*x[0] / 3) * x[0]*x[0] +
+                        x[0] * x[1] +
+                        (-4 + 4 * x[1]*x[1]) * x[1]*x[1];};
+    TasOptimization::ObjectiveFunction shc = TasOptimization::makeObjectiveFunction(num_dimensions, shc_single);
+    state.setNumDimensions(num_dimensions);
+    state.setNumParticles(num_particles);
+    state.initializeParticlesInsideBox(lower, upper);
+    pass = pass & (testParticleSwarmSingle(shc, state, TasDREAM::hypercube(lower, upper), iterations, -1.031628453489877, tol));
+
+    return pass;
+}
+
+} // end namespace
+
+bool DreamExternalTester::testOptimization(){
+    // Tests various optimization states and algorithms
+
+    bool pass1a = TasOptimization::testParticleSwarmState();
+    bool pass1b = TasOptimization::testParticleSwarm();
+
+    return pass1a && pass1b;
```

Comment:
It's good to have separation in tests, but have a test and appropriate `cout` statement that indicates which test failed.

Reply:
Done. I added calls to ``reportPassFail()`` to ``testParticleSwarmState()``, ``testParticleSwarm()``, and ``testOptimization()``.